test/router.js

```diff
@@ -1246,6 +1246,143 @@ describe('Router', function () {
         .expect(200, 'saw GET /bar', done)
     })
   })
+
+  describe('req.matchedRoutes', function () {
+    it('should be set if there is a match', function (done) {
+      var router = new Router()
+      var server = createServer(router)
+      var matchedRoutes
+
+      router.get('/foo', function (req, res, next) {
+        matchedRoutes = req.matchedRoutes
+        next()
+      })
+      router.use(saw)
+
+      request(server)
+          .get('/foo')
+          .expect(200, 'saw GET /foo', function (err, res) {
+            assert.deepEqual(matchedRoutes, ['/foo'])
+            done(err)
+          })
+    })
+
+    it('should be undefined if there is not a match', function (done) {
+      var router = new Router()
+      var server = createServer(router)
+      var matchedRoutes
+
+      router.use(function (req, res, next) {
+        matchedRoutes = req.matchedRoutes
+        next()
+      })
+
+      request(server)
+        .get('/foo')
+        .expect(404, function (err, res) {
+          assert.strictEqual(matchedRoutes, undefined)
+          done(err)
+        })
+    })
+
+    it('should work with sub-routers', function (done) {
+      var router = new Router()
+      var fooRouter = new Router()
+      var server = createServer(router)
+      var matchedFooRoutes
+      var matchedBarRoutes
+
+      router.use('/foo', function (req, res, next) {
+        matchedFooRoutes = req.matchedRoutes
+        next()
+      }, fooRouter)
+      fooRouter.get('/bar', function (req, res, next) {
+        matchedBarRoutes = req.matchedRoutes
+        next()
+      })
+      router.use(saw)
+
+      request(server)
+        .get('/foo/bar')
+        .expect(200, 'saw GET /foo/bar', function (err, res) {
+          assert.deepEqual(matchedFooRoutes, ['/foo'])
+          assert.deepEqual(matchedBarRoutes, ['/foo', '/bar'])
```

Comment:
Do we think it makes sense for the matched foo routes to contain paths that do not even exist on the foo router? What if the foo router had a foo path, would this mean it had a match? That brings up a question for how this works with sub routers mounted at a path that is no / as well.

Reply:
I like to think about the use case where this feature would be used, and if you do not get all matched routes at the end it seems much less useful.  That said, your counter example makes it clear that what we have here is not correct either.